File: lumen-Front/src/api/permissions.ts
/**
 * 权限管理 API 客户端（纯白名单模型）
 */
import type { AclEntry, BatchPermissionRequest } from '../types/permissions';

const API_BASE_URL = 'http://127.0.0.1:8888';

/** GET /permissions/character/{id} */
export async function getCharacterPermissions(
  characterId: string,
  resourceType: string,
  resourceId: string,
): Promise<AclEntry[]> {
  const params = new URLSearchParams({ resource_type: resourceType, resource_id: resourceId });
  const res = await fetch(
    `${API_BASE_URL}/permissions/character/${encodeURIComponent(characterId)}?${params}`,
  );
  if (!res.ok) throw new Error(`获取角色权限失败: ${res.status}`);
  return res.json();
}

/** GET /permissions/resource/{type}/{id} */
export async function getResourcePermissions(
  resourceType: string,
  resourceId: string,
  folderPath: string = '',
  action: string = 'read',
): Promise<string[]> {
  const params = new URLSearchParams({ folder_path: folderPath, action });
  const res = await fetch(
    `${API_BASE_URL}/permissions/resource/${resourceType}/${encodeURIComponent(resourceId)}?${params}`,
  );
  if (!res.ok) throw new Error(`获取资源权限失败: ${res.status}`);
  return res.json();
}

/** POST /permissions/batch-check — 返回 {char_id: boolean} */
export async function batchCheckPermissions(
  resourceType: string,
  resourceId: string,
  folderPath: string,
  characterIds: string[],
  action: string = 'read',
): Promise<Record<string, boolean>> {
  const res = await fetch(`${API_BASE_URL}/permissions/batch-check`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      resource_type: resourceType,
      resource_id: resourceId,
      folder_path: folderPath,
      action,
      character_ids: characterIds,
    }),
  });
  if (!res.ok) throw new Error(`批量权限检查失败: ${res.status}`);
  return res.json();
}

/** PUT /permissions/character/{id}/grant */
export async function grantAccess(
  characterId: string,
  resourceType: string,
  resourceId: string,
  folderPath: string,
  action: string = 'read',
): Promise<{ status: string }> {
  const res = await fetch(`${API_BASE_URL}/permissions/character/${encodeURIComponent(characterId)}/grant`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ resource_type: resourceType, resource_id: resourceId, folder_path: folderPath, action }),
  });
  if (!res.ok) throw new Error(`授权失败: ${res.status}`);
  return res.json();
}

/** DELETE /permissions/character/{id}/revoke */
export async function revokeAccess(
  characterId: string,
  resourceType: string,
  resourceId: string,
  folderPath: string,
  action: string = 'read',
): Promise<{ status: string }> {
  const params = new URLSearchParams({ resource_type: resourceType, resource_id: resourceId, folder_path: folderPath, action });
  const res = await fetch(`${API_BASE_URL}/permissions/character/${encodeURIComponent(characterId)}/revoke?${params}`, {
    method: 'DELETE',
  });
  if (!res.ok) throw new Error(`撤销权限失败: ${res.status}`);
  return res.json();
}

/** PUT /permissions/character/{id} — 批量设置 */
export async function setCharacterPermissions(
  characterId: string,
  data: { resource_type: string; resource_id: string; entries: AclEntry[] },
): Promise<{ status: string; count: number }> {
  const res = await fetch(
    `${API_BASE_URL}/permissions/character/${encodeURIComponent(characterId)}?resource_type=${data.resource_type}&resource_id=${data.resource_id}`,
    { method: 'PUT', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(data.entries) },
  );
  if (!res.ok) throw new Error(`更新角色权限失败: ${res.status}`);
  return res.json();
}

File: api/routes/permissions.py
"""权限管理 API — 纯白名单模型"""
import asyncio
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Literal

router = APIRouter()


class PermissionEntry(BaseModel):
    folder_path: str
    action: Literal["read", "write"] = "read"


class BatchCheckRequest(BaseModel):
    resource_type: str
    resource_id: str
    folder_path: str = ""
    action: Literal["read", "write"] = "read"
    character_ids: List[str]


class GrantRequest(BaseModel):
    resource_type: str
    resource_id: str
    folder_path: str = ""
    action: Literal["read", "write"] = "read"


@router.get("/character/{character_id}")
async def get_character_permissions(
    character_id: str,
    resource_type: str,
    resource_id: str,
):
    """角色视角：获取该角色在某资源下的所有权限规则"""
    from lumen.services.access_control import get_instance
    acl = get_instance()
    return await asyncio.to_thread(
        acl.get_permissions, character_id, resource_type, resource_id
    )


@router.get("/resource/{resource_type}/{resource_id}")
async def get_resource_permissions(
    resource_type: str,
    resource_id: str,
    folder_path: str = "",
    action: str = "read",
):
    """资源视角：获取某路径有权限的角色列表"""
    from lumen.services.access_control import get_instance
    acl = get_instance()
    return await asyncio.to_thread(
        acl.get_characters_with_access, resource_type, resource_id, folder_path, action
    )


@router.post("/batch-check")
async def batch_check_permissions(req: BatchCheckRequest):
    """批量检查多个角色对同一资源的权限，返回 {char_id: bool}"""
    if len(req.character_ids) > 200:
        raise HTTPException(400, "character_ids 超过上限（200）")
    from lumen.services.access_control import get_instance
    acl = get_instance()
    return await asyncio.to_thread(
        acl.batch_check, req.resource_type, req.resource_id,
        req.folder_path, req.action, req.character_ids,
    )


@router.put("/character/{character_id}/grant")
async def grant_access(character_id: str, req: GrantRequest):
    """授予角色访问权限"""
    from lumen.services.access_control import get_instance
    acl = get_instance()
    await asyncio.to_thread(
        acl.grant, character_id, req.resource_type, req.resource_id,
        req.folder_path, req.action,
    )
    return {"status": "ok"}


@router.delete("/character/{character_id}/revoke")
async def revoke_access(
    character_id: str,
    resource_type: str,
    resource_id: str,
    folder_path: str = "",
    action: str = "read",
):
    """撤销角色访问权限"""
    from lumen.services.access_control import get_instance
    acl = get_instance()
    await asyncio.to_thread(
        acl.revoke, character_id, resource_type, resource_id, folder_path, action,
    )
    return {"status": "ok"}


@router.put("/character/{character_id}")
async def set_character_permissions(
    character_id: str,
    resource_type: str,
    resource_id: str,
    entries: List[PermissionEntry],
):
    """批量设置角色的权限规则（覆盖式）"""
    from lumen.services.access_control import get_instance
    acl = get_instance()
    await asyncio.to_thread(
        acl.batch_set, character_id, resource_type, resource_id,
        [e.model_dump() for e in entries]
    )
    return {"status": "ok", "count": len(entries)}
